samples/go/csv/read.go

```diff
@@ -131,24 +169,76 @@ func ReadToMap(r *csv.Reader, structName string, headersRaw []string, pkIdx int,
 			panic(err)
 		}
 
-		var pk types.Value
-		fields := make(types.ValueSlice, len(headersRaw))
-		for i, v := range row {
-			if i < len(headersRaw) {
-				fieldOrigIndex := fieldOrder[i]
-				fields[fieldOrigIndex], err = StringToValue(v, kindMap[fieldOrigIndex])
-				if i == pkIdx {
-					pk = fields[fieldOrigIndex]
-				}
-				if err != nil {
-					d.Chk.Fail(fmt.Sprintf("Error parsing value for column '%s': %s", headersRaw[i], err))
-				}
-			}
-		}
-		kvChan <- pk
+		fields := readFieldsFromRow(row, headersRaw, fieldOrder, kindMap)
+		kvChan <- fields[fieldOrder[pkIndices[0]]]
 		kvChan <- types.NewStructWithType(t, fields)
 	}
+	close(kvChan)
+	return <-mapChan
+}
+
+type mapOrStruct struct {
+	goMap      map[types.Value]mapOrStruct
+	nomsStruct types.Struct
+}
 
+func goMaptoNomsMap(gm map[types.Value]mapOrStruct, vrw types.ValueReadWriter) types.Map {
+	var nomsValue types.Value
+	kvChan := make(chan types.Value, 128)
+	mapChan := types.NewStreamingMap(vrw, kvChan)
+	for k, v := range gm {
+		if v.goMap != nil {
+			nomsValue = goMaptoNomsMap(v.goMap, vrw)
+		} else {
+			nomsValue = v.nomsStruct
+		}
+		kvChan <- k
+		kvChan <- nomsValue
+	}
 	close(kvChan)
 	return <-mapChan
 }
+
+func readToNestedMap(r *csv.Reader, structName string, headersRaw []string, pkIndices []int, t *types.Type, fieldOrder []int, kindMap []types.NomsKind, vrw types.ValueReadWriter) types.Map {
+	goMap := make(map[types.Value]mapOrStruct)
+	for {
+		row, err := r.Read()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			panic(err)
+		}
+
+		fields := readFieldsFromRow(row, headersRaw, fieldOrder, kindMap)
+		rowStruct := types.NewStructWithType(t, fields)
+
+		// needed to allow recursive calls to encloseInMap
+		var encloseInMapFunc func(m map[types.Value]mapOrStruct, keyLevel int) map[types.Value]mapOrStruct
+		encloseInMap := func(m map[types.Value]mapOrStruct, keyLevel int) map[types.Value]mapOrStruct {
+			fieldOrigIndex := fieldOrder[pkIndices[keyLevel]]
```

Comment:
You should be able to say:
var encloseInMap func(...
encloseInMap := func...

there's no need to declare a var with a different name than the one you want to use.